### ISSUE TITLE
Limited support for pad on CoreML backend

### ISSUE THREAD

**Issue by philloooo** [webmachinelearning/webnn#739]
`pad` is [implemented](https://chromium-review.googlesource.com/c/chromium/src/+/5727184) on CoreML backend but with lots of constraints.
On CoreML (See [doc](https://apple.github.io/coremltools/source/coremltools.converters.mil.mil.ops.defs.html#coremltools.converters.mil.mil.ops.defs.iOS15.tensor_operation.pad)), three modes are supported - constant, reflect, replicate(maps to `edge` mode).

1. symmetric mode is not supported (On tflite backend, `edge` mode is not supported.)
2. padding for more than the last two dimensions only supports 'constant' mode. (not documented, but it errors out with this message). One of the webnn samples - [Fast style transfer](https://webmachinelearning.github.io/webnn-samples/style_transfer/) need to use >2D padding for reflect mode.
3. If mode is “reflect” then beginning and ending paddings can be at most input size-1 (same as tensorflow).
4. If mode is “replicate” (aka `edge`) then beginning and ending paddings can be at most input size.

For the third constraint, we can probably update the spec to add such constraint? #377 

For the others, my first question is - @fdwr @huningxin  are there reasonable emulations for these unsupported modes and rank limits?

Further questions are - @mwyrzykowski do you think if CoreML can improve pad support with:
- Add support for `symmetric` mode
- Support padding for more than two dimensions for none `constant` mode.
- Remove the padding value constraints for 'replicate' modes since it doesn't make much sense, it only uses the edge values to replicate.

If none of the restrictions can be emulated / nor supported by CoreML, we will need to either:
a. Add these constraints to WebNN API.
b. Expose the modes limits through `opSupportLimits`. The rank constraints are tricky to expose - as it varies by mode.
c. Just let them fail on CoreML with async error messages.. - Not ideal for browser compatibility.

**Comment by fdwr** [comment]
>  @fdwr @huningxin are there reasonable emulations for these unsupported modes and rank limits?

"Reasonable" is the debatable part, but all of these *can* be emulated (no fewer than 3 operations though). Hopefully future CoreML ops can support them more efficiently someday. Here's pseudocode for each - let me know if anything could make more sense. There are all very similar conceptually to texture wrapping in graphics APIs (OGL, D3D), except in higher dimensions. With the decomposition below, any dimension count is supported, and there are no repetition limits to input size.

## constant

Use `expand` to repeat a constant value, then concatenate the edges along each dimension.

```
result = input
for each axis in input tensor rank
    // If padding present for current dimension on either the low or high end.
    if beginningPadding[axis] != 0 || endingPadding[axis] != 0
        lowChunkDimensions = projectToRank(beginningPadding[axis], input.rank, axis)
        highChunkDimensions = projectToRank(endingPadding[axis], input.rank, axis)

        lowChunk  = expand(scalarTensor, lowChunkDimensions)
        highChunk = expand(scalarTensor, highChunkDimensions)
        result    = concat({lowChunk, result, highChunk}, axis)
    endif
endfor

// projectToRank is a little helper that projects a dimension value up to a given rank at the target axis,
// returning a broadcast-compatible (and concat compatible) a new dimension list.
// e.g. dimension size = 3, rank = 4, axis = 2, output = [1,3,1,1]
```

## edge

```
result = input
for each axis in input tensor rank
   // If padding present for current dimension on either the low or high end.
    if beginningPadding[axis] != 0 || endingPadding[axis] != 0
        dimension = input.dimensions[axis]
        lowChunkStarts      = projectToRank(0, input.rank, axis)
        lowChunkEnds        = projectToRank(1, input.rank, axis)
        highChunkStarts     = projectToRank(dimension - 1, input.rank, axis)
        highChunkEnds       = projectToRank(dimension, input.rank, axis)
        lowChunkDimensions  = projectToRank(beginningPadding[axis], input.rank, axis)
        highChunkDimensions = projectToRank(endingPadding[axis], input.rank, axis)

        lowSlice  = slice(result, starts=lowChunkStarts, ends=lowChunkEnds)
        highSlice = slice(result, starts=highChunkStart, ends=highChunkEnds)
        lowChunk  = expand(lowSlice, lowChunkDimensions)
        highChunk = expand(highSlice, highChunkDimensions)
        result    = concat({lowChunk, result, highChunk}, axis)
    endif
endfor
```

## symmetric

```
mirroredChunk = input
for each axis in input tensor rank
    if beginningPadding[axis] != 0 || endingPadding[axis] != 0
    mirroredChunk = concat({mirroredChunk, reverse(input, axis)}, axis)
    endif
endfor
repetitions  = compute based on ceil(outputDimensions / mirroredChunk size)
outputStarts = compute based on mirroredChunk size, outputDimensions, and beginningPadding
outputEnds   = compute based on mirroredChunk size, outputDimensions, and endingPadding
result       = slice(tile(mirroredChunk, repetitions), outputStarts, outputEnds)
```

## reflection

Same as symmetric, except that you slice off the very edges before mirroring the inner chunk to tile. So [3,4,5] mirrored becomes [3,4,5,4], and then tiled becomes [3,4,5,4,3,4,5,4,3...].

**Comment by lisa0314** [comment]
ORT also can support `symmetric` mode. Maybe we can drop support `symmetric` mode, otherwise we need emulation.

**Comment by lisa0314** [comment]
@fdwr Thanks for point out this typos. Updated `can` to `can't`. Thanks!

**Comment by huningxin** [comment]
@fdwr 

> "Reasonable" is the debatable part, but all of these _can_ be emulated (no fewer than 3 operations though).

I suppose one consideration would be the emulation performance. 

Alternatively, the framework can implement the missing pad mode by WebGPU compute shader, such as ONNX Runtime Web: https://github.com/microsoft/onnxruntime/blob/main/js/web/lib/wasm/jsep/webgpu/ops/pad.ts

@philloooo 
> Expose the modes limits through opSupportLimits

This sounds good to me. We may have something like

```js
dictionary MLPadSupportLimits : MLSingleInputSupportLimits {
  sequence<MLPaddingMode> modes;
};

partial dictionary MLOpSupportLimits {
  MLPadSupportLimits pad;
};
```

**Comment by huningxin** [comment]
Regarding to the discussion in [WebML WG Teleconference – 24 April 2025](https://www.w3.org/2025/04/24-webmachinelearning-minutes.html#cf83), the consensus is to drop the support of "symmetric" padding mode. 

There is another open: the "edge" padding mode is not supported natively by TFLite. It may be emulated, e.g. either by @fdwr 's [solution of using reshape, slice, expand and concat](https://github.com/webmachinelearning/webnn/issues/739#issuecomment-2307958067) or by ["symmetric" padding](https://stackoverflow.com/questions/50174962/tensorflow-how-to-pad-tensor-with-the-edge-values).

It would be interesting to know how many transformers.js models are using "edge" padding mode. @xenova, the data point based https://github.com/huggingface/transformers.js-benchmarking/tree/main/data would be extremely useful.

**Comment by fdwr** [comment]
API             | Constant<br/>x x \|a b c\| x x | Replicate edge<br/>a a \|a b c\| c c | Reflect<br/>c b \|a b c\| b a | Reflect symmetric<br/>b a \|a b c\| c b | Wrap<br/>b c \|a b c\| a b
----------------|------------------------------|------------------------------------|-----------------------------|---------------------------------------|-------------------------------
CoreML          | ✅                          | ✅                                 | ✅                          | ❌                                   | ❌
TF full         | ✅                          | ❌                                 | ✅                          | ✅                                   | ❌
TFLite/LiteRT   | ✅                          | ❌                                 | ✅                          | ✅                                   | ❌
DirectML        | ✅                          | ✅                                 | ✅                          | ✅                                   | ❌
ONNX            | ✅                          | ✅                                 | ✅                          | ❌                                   | ✅
PyTorch         | ✅                          | ✅                                 | ✅                          | ❌                                   | ✅
TOSA            | ✅                          | ❌                                 | ❌                          | ❌                                   | ❌
WebNN 2025-04-29 | ✅                          | ✅                                 | ✅                          | ✅                                   | ❌

- [**CoreML**](https://apple.github.io/coremltools/source/coremltools.converters.mil.mil.ops.defs.html#coremltools.converters.mil.mil.ops.defs.iOS15.tensor_operation.pad) `tensor_operation.pad`: constant, replicate, reflect
- [**DirectML**](https://learn.microsoft.com/en-us/windows/win32/api/directml/ne-directml-dml_padding_mode) `DML_OPERATOR_PADDING`: constant, edge, reflection, symmetric
- [**ONNX**](https://onnx.ai/onnx/operators/onnx__Pad.html) `Pad`: constant, edge, reflect, wrap
- [**PyTorch**](https://pytorch.org/docs/stable/generated/torch.nn.functional.pad.html) `torch.nn.functional.pad`: constant, replicate, reflect, circular
- [**TOSA**](https://mlir.llvm.org/docs/Dialects/TOSA/#tosapad-mlirtosapadop) `tosa.pad`: const
- [**TF**](https://www.tensorflow.org/api_docs/python/tf/pad) `tf.pad`: constant, reflect, symmetric
- [**TFLite**](https://www.tensorflow.org/mlir/tfl_ops#tflmirror_pad_tflmirrorpadop) [`tfl.pad`](https://www.tensorflow.org/mlir/tfl_ops#tflpad_tflpadop) and [`tfl.mirror_pad`](https://www.tensorflow.org/mlir/tfl_ops#tflmirror_pad_tflmirrorpadop): constant, reflect, symmetric
- [**WebNN**](https://webmachinelearning.github.io/webnn/#enumdef-mlpaddingmode) `pad`: constant, edge, reflection, symmetric

**Comment by huningxin** [comment]
@xenova , thanks so much, the tool is so useful!

> of ~1250 unique model ids (~5500 onnx files), it only appears like the Mimi audio encoder uses edge padding.

And it looks like there is only one edge pad in Mimi audio encoder, all others are constant pad.

**Comment by xenova** [comment]
I've now updated the list to ~1600 different models ids (7700 onnx files), and still, only the Mimi encoder uses this kind of padding. And as @huningxin points out, it's only used in a single node.

**Comment by fdwr** [comment]
> I made some improvements to my collection process (https://github.com/huggingface/transformers.js-benchmarking/pull/5), and of ~1250 unique model ids (~5500 onnx files), it only appears like the Mimi audio encoder uses edge padding. ... created an interactive model explorer which I hope can be helpful!

@xenova : Cool. Thank you for the model analysis and sharing the model explorer (convenient not need to locally download the model file first to use Netron).

> After overriding the free dimensions, netron shows this edge pad pads the last dimension from 1 to 4 with [2, 1] padding size.

@huningxin So then the Mimi case where edge replication is used, it pads with a size that is *greater* than the original dimension length, which makes it more important to avoid size restrictions for that mode (and Joshua already applied in his [latest commit](https://github.com/webmachinelearning/webnn/pull/843/files/9129a11ad4eb59c51d20236d967374b95d7b9086..79da15b94ed5afcd39030a0b415fe8c622d358a9).
